pkg/spoke/managedclusteraddon/lease_controller.go

```diff
@@ -0,0 +1,182 @@
+	if errors.IsNotFound(err) {
+		// addon is not found, could be deleted, ignore it.
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	return c.syncSingle(ctx, addOnInstallNamespace, addOn, syncCtx.Recorder())
+}
+
+func (c *managedClusterAddOnLeaseController) syncSingle(ctx context.Context,
+	leaseNamespace string,
+	addOn *addonapiv1alpha1.ManagedClusterAddOn,
+	recorder events.Recorder) error {
+	var condition metav1.Condition
+	//TODO: add this to addon spec or export this as a flag
+	leaseDurationSeconds := 60
+	gracePeriod := time.Duration(leaseDurationTimes*leaseDurationSeconds) * time.Second
+
+	observedLease, err := c.leaseLister.Leases(leaseNamespace).Get(addOn.Name)
+	switch {
+	case errors.IsNotFound(err):
+		condition = metav1.Condition{
+			Type:    "ManagedClusterAddOnConditionAvailable", //TODO: add this to api
+			Status:  metav1.ConditionUnknown,
+			Reason:  "ManagedClusterAddOnLeaseNotFound",
+			Message: "Managed cluster addon agent lease is not found.",
+		}
+	case err != nil:
+		return err
+	case err == nil:
+		if time.Now().Before(observedLease.Spec.RenewTime.Add(gracePeriod)) {
+			// the lease is constantly updated, update its addon status to available
+			condition = metav1.Condition{
+				Type:    "ManagedClusterAddOnConditionAvailable", //TODO: add this to api
+				Status:  metav1.ConditionTrue,
+				Reason:  "ManagedClusterAddOnLeaseUpdated",
+				Message: "Managed cluster addon agent updates its lease constantly.",
+			}
+			break
+		}
+
+		// the lease is not constantly updated, update its addon status to unavailable
+		condition = metav1.Condition{
+			Type:    "ManagedClusterAddOnConditionAvailable", //TODO: add this to api
+			Status:  metav1.ConditionFalse,
+			Reason:  "ManagedClusterAddOnLeaseUpdateStopped",
+			Message: "Managed cluster addon agent stopped updating its lease.",
+		}
+	}
+
+	if meta.IsStatusConditionPresentAndEqual(addOn.Status.Conditions, condition.Type, condition.Status) {
+		// addon status is not changed, do nothing
+		return nil
+	}
+
+	_, updated, err := helpers.UpdateManagedClusterAddOnStatus(
+		ctx,
+		c.addOnClient,
+		c.clusterName,
+		addOn.Name,
+		helpers.UpdateManagedClusterAddOnStatusFn(condition),
+	)
+	if err != nil {
+		return err
+	}
+	if updated {
+		recorder.Eventf("ManagedClusterAvailableConditionUpdated",
+			"update managed cluster addon %q available condition to unavailable, due to its lease is not updated constantly",
+			addOn.Name)
+	}
+
+	return nil
+}
+
+func (c *managedClusterAddOnLeaseController) queueKeyFunc(lease runtime.Object) string {
+	accessor, _ := meta.Accessor(lease)
+
+	name := accessor.GetName()
+	// addon lease name should be same with the lease name.
+	addOn, err := c.addOnLister.ManagedClusterAddOns(c.clusterName).Get(name)
+	if err != nil {
+		// failed to get addon from hub, ignore this reconciliation.
+		return ""
+	}
+
+	addOnInstallNamespace, ok := addOn.Annotations["installNamespace"]
```

Comment:
we need a common func to extract installNamespace and maybe others together.

Reply:
I will make a common func for this